MIPs/mip-3.md

```diff
@@ -0,0 +1,184 @@
+
+- `from`: An object containing details about the source token. It should include:
+
+  - `tokenAddress`: The address of the source token.
+  - `value`: The amount of wei in hexadecimal format of the source token to be swapped.
+
+- `to`: An object containing details about the destination token. It should include:
+
+  - `tokenAddress`: The address of the destination token.
+
+- `userAddress`: Account address connected to the dapp.
+
+- `sendTo`: (Optional) Allows integration with Swap's `send to` feature.
+
+MetaMask will interpret the method call and perform the necessary validations and operations to initiate the token swap.
+
+## Implementation
+
+The MetaMask team will be responsible for implementing the JSON RPC `wallet_swapAsset` method. This will involve updating the wallet's internal logic.
+
+- PR for MetaMask mobile: https://github.com/MetaMask/metamask-mobile/pull/7509
+
+## Developer Adoption Considerations
+
+Considering that this is a new json rpc method, developers should take the following aspects into account when adopting this proposal:
+
+1. Backward compatibility: The introduction of the `wallet_swapAsset` method does not affect existing methods or functionality. Developers can integrate this method into their dApps without needing to modify existing code.
+
+2. Ease of adoption: The `wallet_swapAsset` method is designed to be straightforward to use. Developers simply need to provide the necessary parameters (source token details, destination token details) in the method call. The wallet handles the rest of the process, including validation, token swap operation, and error handling.
+
+3. User Experience: By using the `wallet_swapAsset` method, developers can improve the user experience of their dApps. Users can perform token swaps without leaving the dApp, making the process more seamless and convenient.
+
+4. Error handling: Dapp Developers should implement appropriate error handling when using the `wallet_swapAsset` method. The method can return various errors, such as invalid parameters or unsupported operations, and these should be handled gracefully in the dApp.
+   On MetaMask side we intent to validate every property of the request and return an approriate answer to the dapp.
+   Currently we throw these errors:
+
+5. Ensuring Correct Chain Context for wallet_swapAsset: Before initiating a swap, it's crucial to ensure that the user's wallet is connected to the expected blockchain network. Developers should use the [eth_chainId](https://docs.metamask.io/wallet/reference/eth_chainid/) method to verify the current chain, [wallet_switchEthereumChain](https://docs.metamask.io/wallet/reference/wallet_switchethereumchain/) to switch to the desired chain if necessary, and [wallet_addEthereumChain](https://docs.metamask.io/wallet/reference/wallet_addethereumchain/) to add a new chain if it's not known to the wallet. This ensures that token swaps are performed on the intended network, enhancing security and user experience.
```

Comment:
```suggestion
5. Ensuring Correct Chain Context for `wallet_swapAsset`: Before initiating a swap, it's crucial to ensure that the user's wallet is connected to the expected blockchain network. Developers should use the [eth_chainId](https://docs.metamask.io/wallet/reference/eth_chainid/) method to verify the current chain, [wallet_switchEthereumChain](https://docs.metamask.io/wallet/reference/wallet_switchethereumchain/) to switch to the desired chain if necessary, and [wallet_addEthereumChain](https://docs.metamask.io/wallet/reference/wallet_addethereumchain/) to add a new chain if it's not known to the wallet. This ensures that token swaps are performed on the intended network, enhancing security and user experience.
```

Reply:
Addressed on this [commit](https://github.com/MetaMask/metamask-improvement-proposals/pull/25/commits/d64432f9dc185bbd80153fa1950aa18ee8b74db7)